Documentation/teaching/lectures/smp.rst

```diff
@@ -0,0 +1,1156 @@
+     +-------------+               | cEEE       |               +-------------+
+                                   +------------+
+
+
+In order to provide atomic operations on SMP systems different
+architectures use different techniques. For example, on x86 a LOCK
+prefix is used to lock the system bus while executing the prefixed
+operation:
+
+.. slide:: Fixing atomic operations for SMP systems (x86)
+   :inline-contents: True
+   :level: 2
+
+   |_|
+
+   .. ditaa::
+
+                                   +------------+
+     +-------------+   BUS LOCK    |   Memory   |
+     |    CPU 1    |<------------->|            |
+     |             |   LOAD (0)    |            |
+     |    inc v    |<--------------|   v <- 0   |
+     |             |   STORE (1)   |            |
+     |             |-------------->|   v <- 1   |
+     |             |  BUS UNLOCK   |            |
+     | cEEE        |<------------->|            |   BUS LOCK    +-------------+
+     +-------------+               |            |<------------->|    CPU 1    |
+                                   |            |   LOAD (1)    |             |
+                                   |            |<--------------|    inc v    |
+                                   |   v <- 2   |   STORE (2)   |             |
+                                   |            |-------------->|             |
+                                   |            |  BUS UNLOCK   |             |
+                                   | cEEE       |<------------->| cEEE        |
+                                   +------------+               +-------------+
+
+
+On ARM the LDREX and STREX instructions are used together to guarantee
+atomic access: LDREX loads a value and signals the exclusive monitor
+that an atomic operation is in progress. The STREX attempts to store a
+new value but only succeeds if the exclusive monitor has not detected
+other exclusive operations. So, to implement atomic operations the
+programmer must retry the operation (both LDREX and STREX) until the
+exclusive monitor signals a success.
+
+As seen from the implementation details, atomic operations are
+expensive and they must be avoided when possible.
+
+Preemption (interrupts) disabling
+=================================
+
+On single core systems and non preemptive kernels the only source of
+concurrency is the preemption of the current thread by an
+interrupt. To prevent concurrency is thus sufficient to disable
+interrupts.
+
+This is done with architecture specific instructions, but Linux offers
+architecture independent APIs to disable and enable interrupts:
+
+.. slide:: Synchronization with interrupts (x86)
+   :inline-contents: True
+   :level: 2
+
+   .. code-block:: c
+
+       #define local_irq_disable() \
+           asm volatile („cli” : : : „memory”)
+
+      #define local_irq_enable() \
+          asm volatile („sti” : : : „memory”)
+
+      #define local_irq_save(flags) \
```

Comment:
Here it appears that interrupts are not disabled, but they are, see the code: https://elixir.bootlin.com/linux/v4.15/source/arch/x86/include/asm/irqflags.h#L108
static inline notrace unsigned long arch_local_irq_save(void)
{
	unsigned long flags = arch_local_save_flags();
	arch_local_irq_disable();
	return flags;
}